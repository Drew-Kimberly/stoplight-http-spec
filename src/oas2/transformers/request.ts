import { IHttpOperationRequest } from '@stoplight/types';
import { BodyParameter, FormDataParameter, Parameter } from 'swagger-schema-official';

import { isBodyParameter, isFormDataParameter, isHeaderParameter, isPathParameter, isQueryParameter } from '../guards';
import {
  translateFromFormDataParameter,
  translateToBodyParameter,
  translateToHeaderParam,
  translateToPathParameter,
  translateToQueryParameter,
} from './params';

export function translateToRequest(parameters: Parameter[], consumes: string[]): IHttpOperationRequest {
  const bodyParameters = parameters.filter(parameter => isBodyParameter(parameter)) as BodyParameter[];
  const formDataParameters = parameters.filter(parameter => isFormDataParameter(parameter)) as FormDataParameter[];
  const request: IHttpOperationRequest = {};

  // if 'body' and 'form data' defined prefer 'body'
  if (!!bodyParameters.length) {
    // There can be only one body parameter (taking first one)
    request.body = translateToBodyParameter(bodyParameters[0], consumes);
  } else if (!!formDataParameters.length) {
    request.body = translateFromFormDataParameter(formDataParameters, consumes);
  }

  return parameters.reduce(reduceRemainingParameters, request);
}

function reduceRemainingParameters(request: IHttpOperationRequest, parameter: Parameter) {
  if (isQueryParameter(parameter)) {
    const queryParameter = translateToQueryParameter(parameter);
    request.query = (request.query || []).concat(queryParameter);
  } else if (isPathParameter(parameter)) {
    const pathParameter = translateToPathParameter(parameter);
    request.path = (request.path || []).concat(pathParameter);
  } else if (isHeaderParameter(parameter)) {
    const headerParameter = translateToHeaderParam(parameter);
    request.headers = (request.headers || []).concat(headerParameter);
  }
  return request;
}
