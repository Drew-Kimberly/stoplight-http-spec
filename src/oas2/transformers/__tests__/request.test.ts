import { FormDataParameter } from 'swagger-schema-official';

import {
  isBodyParameter,
  isFormDataParameter,
  isHeaderParameter,
  isPathParameter,
  isQueryParameter,
} from '../../guards';
import {
  translateFromFormDataParameter,
  translateToBodyParameter,
  translateToHeaderParam,
  translateToPathParameter,
  translateToQueryParameter,
} from '../params';
import { translateToRequest } from '../request';

jest.mock('../params');
jest.mock('../../guards');

describe('request', () => {
  const consumes = ['*'];
  const fakeParameter: FormDataParameter = {
    name: 'name',
    type: 'string',
    in: 'formData',
  };
  const fakeBodyParameter = { in: 'body' };
  const fakeFormParameter = { in: 'body', type: 'form' };
  const fakeQueryParameter = { in: 'query' };
  const fakeHeaderParameter = { in: 'header' };
  const fakePathParameter = { in: 'path' };

  beforeEach(() => {
    (translateToBodyParameter as jest.Mock).mockReturnValue(fakeBodyParameter);
    (translateFromFormDataParameter as jest.Mock).mockReturnValue(fakeFormParameter);
    (translateToQueryParameter as jest.Mock).mockReturnValue(fakeQueryParameter);
    (translateToPathParameter as jest.Mock).mockReturnValue(fakePathParameter);
    (translateToHeaderParam as jest.Mock).mockReturnValue(fakeHeaderParameter);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  test('given empty params collection should return empty object', () => {
    expect(translateToRequest([], consumes)).toEqual({});
  });

  test('given single body param should translate to request with body', () => {
    (isBodyParameter as any).mockReturnValue(true);
    expect(translateToRequest([fakeParameter], consumes)).toMatchSnapshot();
  });

  test('given single form param should translate to request with form', () => {
    (isFormDataParameter as any).mockReturnValue(true);
    expect(translateToRequest([fakeParameter], consumes)).toMatchSnapshot();
  });

  test('given single path param should translate to request with path', () => {
    (isPathParameter as any).mockReturnValue(true);
    expect(translateToRequest([fakeParameter], consumes)).toMatchSnapshot();
  });

  test('given single query param should translate to request with query', () => {
    (isQueryParameter as any).mockReturnValue(true);
    expect(translateToRequest([fakeParameter], consumes)).toMatchSnapshot();
  });

  test('given single header param should translate to request with header', () => {
    (isHeaderParameter as any).mockReturnValue(true);
    expect(translateToRequest([fakeParameter], consumes)).toMatchSnapshot();
  });

  test('given single unknown param should translate to empty request', () => {
    expect(translateToRequest([fakeParameter], consumes)).toEqual({});
  });

  test('given two query params should translate', () => {
    (isQueryParameter as any).mockReturnValueOnce(true);
    (isQueryParameter as any).mockReturnValueOnce(true);
    expect(translateToRequest([fakeParameter, fakeParameter], consumes)).toMatchSnapshot();
  });

  test('given two header params should translate', () => {
    (isHeaderParameter as any).mockReturnValueOnce(true);
    (isHeaderParameter as any).mockReturnValueOnce(true);
    expect(translateToRequest([fakeParameter, fakeParameter], consumes)).toMatchSnapshot();
  });

  test('given two path params should translate', () => {
    (isPathParameter as any).mockReturnValueOnce(true);
    (isPathParameter as any).mockReturnValueOnce(true);
    expect(translateToRequest([fakeParameter, fakeParameter], consumes)).toMatchSnapshot();
  });

  test('should translate mixed request', () => {
    (isBodyParameter as any).mockReturnValueOnce(true);
    (isFormDataParameter as any).mockReturnValueOnce(true);
    (isQueryParameter as any).mockReturnValueOnce(true);
    (isHeaderParameter as any).mockReturnValueOnce(true);
    (isPathParameter as any).mockReturnValueOnce(true);
    expect(
      translateToRequest([fakeParameter, fakeParameter, fakeParameter, fakeParameter, fakeParameter], consumes),
    ).toMatchSnapshot();
  });
});
