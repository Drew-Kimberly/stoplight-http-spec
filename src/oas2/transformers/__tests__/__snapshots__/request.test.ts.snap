// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`request given single body param should translate to request with body 1`] = `
Object {
  "body": Object {
    "in": "body",
  },
}
`;

exports[`request given single form param should translate to request with form 1`] = `
Object {
  "body": Object {
    "in": "body",
    "type": "form",
  },
}
`;

exports[`request given single header param should translate to request with header 1`] = `
Object {
  "headers": Array [
    Object {
      "in": "header",
    },
  ],
}
`;

exports[`request given single path param should translate to request with path 1`] = `
Object {
  "path": Array [
    Object {
      "in": "path",
    },
  ],
}
`;

exports[`request given single query param should translate to request with query 1`] = `
Object {
  "query": Array [
    Object {
      "in": "query",
    },
  ],
}
`;

exports[`request given two header params should translate 1`] = `
Object {
  "headers": Array [
    Object {
      "in": "header",
    },
    Object {
      "in": "header",
    },
  ],
}
`;

exports[`request given two path params should translate 1`] = `
Object {
  "path": Array [
    Object {
      "in": "path",
    },
    Object {
      "in": "path",
    },
  ],
}
`;

exports[`request given two query params should translate 1`] = `
Object {
  "query": Array [
    Object {
      "in": "query",
    },
    Object {
      "in": "query",
    },
  ],
}
`;

exports[`request should translate mixed request 1`] = `
Object {
  "body": Object {
    "in": "body",
  },
  "headers": Array [
    Object {
      "in": "header",
    },
  ],
  "path": Array [
    Object {
      "in": "path",
    },
  ],
  "query": Array [
    Object {
      "in": "query",
    },
  ],
}
`;
